src/nodedumper.h

```diff
@@ -1,39 +1,65 @@
 #pragma once
 
 #include <string>
-#include <map>
+#include <unordered_map>
 #include <list>
 #include "NodeVisitor.h"
 #include "node.h"
 #include "nodecache.h"
 
+// GroupNodeChecker does a quick first pass to count children of group nodes
+// If a GroupNode has 0 children, don't include in node id strings
+// If a GroupNode has 1 child, we replace it with its child
+// This makes id strings much more compact for deeply nested trees, recursive scad scripts,
+// and increases likelihood of node cache hits.
+class GroupNodeChecker : public NodeVisitor 
+{
+public:
+    GroupNodeChecker(){}
+
+    Response visit(State &state, const AbstractNode &node) override;
+    Response visit(State &state, const GroupNode &node) override;
+    //Response visit(State &state, const RootNode &node) override;
+    void incChildCount(int groupNodeIndex);
+    int getChildCount(int groupNodeIndex);
+    void reset() { groupChildCounts.clear(); }
+
+private:
+    // stores <node_idx,nonEmptyChildCount> for each group node
+    std::unordered_map<int, int> groupChildCounts;
+};
 
 class NodeDumper : public NodeVisitor
 {
 public:
-        /*! If idPrefix is true, we will output "n<id>:" in front of each node,
-          which is useful for debugging. */
-        NodeDumper(NodeCache &cache, const AbstractNode *root_node, const std::string& indent, bool idString, bool idPrefix) :
-                cache(cache), indent(indent), idString(idString), idprefix(idPrefix), currindent(0), root(root_node) { }
-        ~NodeDumper() {}
+    NodeDumper(NodeCache &cache, const AbstractNode *root_node, const std::string& indent, bool idString) :
+            cache(cache), indent(indent), idString(idString), currindent(0), root(root_node) { 
+        if (idString) { 
+            groupChecker.reset();
```

Comment:
Since groupChecker gets constructed here as well. it looks like this reset() call is redundant